lib/streamlit/delta_generator.py

```diff
@@ -908,35 +931,26 @@ def _melt_data(df: DataFrame, last_index: Any) -> tuple[DataFrame, int | Any]:
                     "'RangeIndex' object has no attribute 'step'"
                 )
 
-            start = last_index + old_step
-            stop = last_index + old_step + old_stop
+            start = add_rows_metadata.last_index + old_step
+            stop = add_rows_metadata.last_index + old_step + old_stop
 
             df.index = pd.RangeIndex(start=start, stop=stop, step=old_step)
-            last_index = stop - 1
+            add_rows_metadata.last_index = stop - 1
 
-        index_name = df.index.name
-        if index_name is None:
-            index_name = "index"
+        if is_legacy:
+            index_name = df.index.name
+            if index_name is None:
+                index_name = "index"
 
-        df = pd.melt(df.reset_index(), id_vars=[index_name])
-        return df, last_index
-
-    # For some delta types we have to reshape the data structure
-    # otherwise the input data and the actual data used
-    # by vega_lite will be different, and it will throw an error.
-    if (
-        delta_type in DELTA_TYPES_THAT_MELT_DATAFRAMES
-        or delta_type in ARROW_DELTA_TYPES_THAT_MELT_DATAFRAMES
-    ):
-        if not isinstance(data, pd.DataFrame):
-            return _melt_data(
-                df=type_util.convert_anything_to_df(data),
-                last_index=last_index,
-            )
+            out_data = pd.melt(df.reset_index(), id_vars=[index_name])
         else:
-            return _melt_data(df=data, last_index=last_index)
+            out_data, *_ = prep_data(df, **add_rows_metadata.columns)
```

Comment:
Properly handle the data add_rows by doing to it the exact same things we do in _generate_chart()